system-tests-new/incremental/test_incremental.py

```diff
@@ -0,0 +1,225 @@
+                if 'PRINT' in line:
+                    out.write(line)
+
+        # rebuild
+        rebuild_files, rebuild_timestamps, rebuild_hashes = self.build(build_config)
+
+        # ensure the analysis and object files change but the mod file does not
+        prebuild_files = filter(lambda p: PREBUILD in str(p), rebuild_files)
+        prebuild_groups = get_prebuild_file_groups(prebuild_files)
+        prebuild_folder = build_config.prebuild_folder
+
+        # my_prog should be completely unaffected
+        self.assert_one_artefact(
+            ['my_prog.*.an', 'my_prog.*.o'],
+            prebuild_groups, prebuild_folder, clean_timestamps, clean_hashes, rebuild_timestamps, rebuild_hashes)
+
+        # my_mod will have a new mod file because the source has changed, so it's recompiled into a different artefact,
+        # but the interface hasn't changed so we expect the mod contents to be identical.
+        self.assert_two_identical_artefacts(
+            ['my_mod.*.mod'],
+            prebuild_groups, prebuild_folder, rebuild_hashes)
+
+        self.assert_two_different_artefacts(
+            ['my_mod.*.an', 'my_mod.*.o'],
+            prebuild_groups, prebuild_folder, rebuild_hashes)
+
+    def test_mod_interface_change(self, build_config):
+        # test a module interface change
+
+        # clean build
+        clean_files, clean_timestamps, clean_hashes = self.build(build_config)
+
+        # modify the fortran module source, changing the module interface
+        mod_source = build_config.source_root / 'src/my_mod.F90'
+        lines = open(mod_source, 'rt').readlines()
+        with open(mod_source, 'wt') as out:
+            for line in lines:
+                out.write(line)
+                # add an extra subroutine
+                if 'END SUBROUTINE' in line:
+                    out.write("""
+                        SUBROUTINE added_func ()
+
+                            INTEGER :: bar = 2
+                            PRINT *, bar
+
+                        END SUBROUTINE added_func
+                    """)
+
+        # rebuild
+        rebuild_files, rebuild_timestamps, rebuild_hashes = self.build(build_config)
+
+        # ensure the analysis and object files change but the mod file does not
+        prebuild_files = filter(lambda p: PREBUILD in str(p), rebuild_files)
+        prebuild_groups = get_prebuild_file_groups(prebuild_files)
+        prebuild_folder = build_config.prebuild_folder
+
+        # my_prog analysis should be unaffected
+        self.assert_one_artefact(
+            ['my_prog.*.an'],
+            prebuild_groups, prebuild_folder, clean_timestamps, clean_hashes, rebuild_timestamps, rebuild_hashes)
+
+        # We've recompiled my_prog because a mod it depends on changed.
+        # That means there'll be a different version of the artefact with a new hash of things it depends on.
+        # However, it's not *doing* anything different, so the object file should have the same contents.
```

Comment:
Presumably there is no change because although a new routine has been added to the API it isn't called. Please clarify this for future developers.

Reply:
Good point.